### ISSUE TITLE
Explore async Rust runtimes

### ISSUE THREAD

**Issue by sergefdrv** [replica-io/replica-io#70]
Explore the structure and techniques in async Rust runtimes: [Tokio](https://tokio.rs/), [async-std](https://async.rs/), and [smol](https://github.com/smol-rs/smol).

**Comment by anchalshivank** [comment]
Do check out glossary at  #55 
### Understanding How a Runtime Works in Rust

#### Why Do We Need a Runtime?

In many programming languages like Java or Python, asynchronous operations are straightforward because these languages provide a built-in runtime to manage asynchronous tasks. However, Rust does not provide an official default runtime for asynchronous programming. This means that when writing asynchronous code in Rust, you need to rely on third-party runtimes like **Tokio**, **async-std**, or **smol**.

#### What is a Runtime?

A **runtime** is an environment that manages the execution of asynchronous tasks, particularly I/O-bound tasks. Without a runtime, the system wouldn't know how to manage tasks that can run concurrently without blocking the main thread. The runtime takes care of task scheduling, handling I/O operations, and ensuring that tasks are executed efficiently.

#### A Real-World Analogy

Imagine you have a worker, WorkerA, who is tasked with packaging parcels. Packaging depends on wrappers being available, so WorkerA often needs to wait for the wrappers to arrive. Instead of sitting idle, WorkerA can optimize their time by taking on another task while waiting for the wrappers.

Similarly, for a web server, if **Request A** depends on data from a database, instead of waiting idly for the database response, the server can handle **Request B** in the meantime. This way, resources are utilized efficiently without wasting time waiting.

#### The Problem with Synchronous Code

By default, code runs **synchronously**: it executes one line at a time, from top to bottom, blocking further execution until each operation completes. For I/O-bound tasks, this behavior isn't ideal because it wastes time waiting for resources like disk access or network responses. To solve this, we make the code **asynchronous** to handle multiple tasks more efficiently.

#### Core Components of a Runtime

To understand how asynchronous runtimes work in Rust (like Tokio, async-std, and smol), let’s break down the essential components of a runtime:

1. **Executor**: Responsible for **polling and scheduling tasks**. It manages the execution of asynchronous tasks by running them when they are ready to proceed.
   
2. **Reactor**: Works with the operating system to **monitor I/O events**. It is responsible for handling I/O-bound tasks like reading from the network or a file system without blocking the thread.
   
3. **Task**: The smallest unit of work that can be executed. In Rust, these are usually represented by `Future`s, which are lazy and need to be **polled** to produce a result.

4. **Scheduler**: Often overlooked, the scheduler is part of the executor. It determines **which tasks to run and when**. The scheduler makes decisions based on task readiness and availability of resources, ensuring fairness and efficiency in task execution.

#### Code Example: Asynchronous Functions

```rust
async fn taskA() -> i32 {
    // Simulate a delay
    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
    5
}

async fn taskB() -> i32 {
    tokio::time::sleep(std::time::Duration::from_secs(2)).await;
    2
}

#[tokio::main] // This macro sets up the Tokio runtime
async fn main() {
    let a = taskA().await;
    let b = taskB().await;

    println!("{}", a + b); // Prints 7 after both tasks complete
}
```

### Breaking Down the Execution

- When the code encounters the `await` keyword inside `taskA()` and `taskB()`, the functions return **futures**. Futures in Rust are **lazy**, meaning they do not start execution until they are **polled** by the runtime.

- The runtime stores these tasks (represented by futures) in a **task queue**.

- The **reactor** interacts with the OS to register I/O operations. For example, when `sleep` is called, it doesn't block the thread. Instead, the reactor registers a timer with the OS, asking to be notified once the timer expires.

- While waiting for I/O operations (like the timer), the **scheduler** in the executor can switch to another task that’s ready to run, allowing the system to make progress on multiple tasks.

### Event Loop and Task Polling

At the heart of every runtime is the **event loop**, a loop that continuously checks for tasks that are ready to be executed. The event loop operates as follows:

1. **Poll tasks**: The executor polls tasks in its queue to check if they are ready to continue execution.
   
2. **Reactor event handling**: The reactor listens for events (e.g., I/O completion) from the OS. Once an event occurs, it signals the executor that a task is ready to proceed.

3. **Wakeup and polling**: When an I/O operation completes, a **wakeup** signal is sent to the executor. The executor will then poll the future again to resume its execution until the future completes.

This is a **cooperative multitasking** system, meaning that tasks voluntarily yield control when they are waiting on I/O. This allows the runtime to handle many tasks efficiently without blocking threads unnecessarily.

#### Corrected Explanation of Executor and Reactor

- **Executor**: Manages the lifecycle of tasks. It schedules when tasks should be polled and runs them when they are ready. The executor ensures that tasks which yield (via `await`) are suspended and then resumed when they can make progress.

- **Reactor**: This component works with the OS to handle I/O events. It doesn’t execute the tasks itself but **registers I/O events** (like timers or network I/O) with the OS. When the OS signals that an event is ready, the reactor notifies the executor to resume the relevant task.

### Conclusion

Runtimes in Rust, like Tokio and async-std, provide the necessary infrastructure to handle asynchronous tasks without blocking the main thread. The **executor**, **reactor**, and **scheduler** work together to manage tasks efficiently by taking advantage of asynchronous I/O and cooperative multitasking. Writing asynchronous code in Rust allows us to optimize performance, especially in I/O-bound applications like web servers, without blocking threads unnecessarily.

**Comment by anchalshivank** [comment]
Let's dive into the need for multiple asynchronous runtimes in Rust and compare `Tokio`, `async-std`, and `smol`.

### Popular runtimes

- **Tokio** came first and established a robust asynchronous runtime for Rust, but it was complex, particularly with its distinction between single-threaded and multi-threaded runtimes. Additionally, its ecosystem was somewhat fragmented, requiring users to opt into various components manually.
  
- **async-std** was created to offer a simpler and more ergonomic experience, aiming for closer alignment with Rust’s `std` library and to follow an interface inspired by JavaScript's async/await and Go’s concurrency model.
  
- **smol** later emerged as a lightweight alternative focusing on minimalism and performance, designed to be fast and easy to integrate without heavyweight dependencies.

### Core Components of an Async Runtime

All these runtimes share some common components, though they differ in implementation and philosophy. The essential parts of any async runtime include:

1. **Reactor**: Handles the readiness of I/O operations. In an async context, tasks yield when waiting for an I/O operation and the reactor wakes them up once the operation is ready to resume.

2. **Executor**: Manages the execution of tasks. It schedules tasks, allows them to yield, and resumes them when needed.

3. **Scheduler**: Balances tasks across threads (if multi-threaded) or within a single-threaded environment.

4. **Task**: The smallest unit of work that can be scheduled. These are futures in Rust's async programming.

### Comparison of `Tokio`, `async-std`, and `smol`

| Feature/Component    | **Tokio**                                         | **async-std**                                 | **smol**                                         |
|----------------------|--------------------------------------------------|----------------------------------------------|--------------------------------------------------|
| **Philosophy**        | Performance-first, feature-rich runtime          | Simplicity and alignment with `std` APIs     | Minimal, lightweight, aims for a small footprint |
| **Reactor**           | Uses `mio` (Multi-threaded I/O reactor based on epoll, kqueue, etc.) | Based on `mio` but more abstracted for simplicity | Tiny reactor built on `polling` crate             |
| **Executor**          | Multi-threaded by default (optional single-threaded runtime) | Single-threaded by default, but supports multi-threaded | Very lightweight single-threaded executor with thread-pool option |
| **Scheduler**         | **Work-Stealing Scheduler**: Tries to balance tasks across multiple threads for optimal performance | Simpler **Round-Robin Scheduler** for fairness but less optimized | Lightweight task stealing if using thread-pool or simple round-robin |
| **Ecosystem**         | Largest ecosystem with many utilities (e.g., `tokio::io`, `tokio::net`, etc.) | Smaller, but similar to `std::` interfaces   | Minimal ecosystem focused on async primitives   |
| **Concurrency**       | High concurrency support with a focus on performance and scalability | Good concurrency support but with simplicity in mind | Focuses on lightweight task scheduling          |
| **Ease of Use**       | Complex; requires feature flags and configuration | Easier to use, more like `std`                | Easiest, with minimal setup and boilerplate      |
| **Integration**       | Requires libraries specifically built for Tokio's async features (`tokio::net`, `tokio::fs`, etc.) | Can work well with `std::future` and `async` primitives | Very simple integration, more lightweight        |
| **Performance**       | Top-tier performance for large-scale, I/O-heavy systems | Moderate performance with an ergonomic design | Designed for speed but targets smaller workloads |
| **Memory Usage**      | More complex memory management, especially with larger workloads | Simpler memory management, though can increase with multithreading | Extremely low memory overhead                   |

### Internal Algorithms and Components

#### 1. **Tokio**

- **Reactor**: Based on `mio`, a cross-platform library for building high-performance I/O applications. It uses the OS’s native I/O event notification systems (`epoll` on Linux, `kqueue` on macOS, etc.) to manage readiness and wake tasks.
  
- **Executor**: **Work-Stealing Scheduler**. Tokio uses a work-stealing algorithm to balance the load between threads, which is efficient for high concurrency workloads. This approach allows threads with fewer tasks to "steal" work from busier threads, leading to optimal CPU utilization.

- **Task**: Tokio’s tasks are future-based, and the runtime ensures efficient suspension and resumption of these tasks without unnecessary thread context switching.

#### 2. **async-std**

- **Reactor**: Also based on `mio`, but abstracted to be simpler for users, mimicking the feel of synchronous Rust `std` APIs but operating asynchronously.

- **Executor**: **Round-Robin Scheduler**. Tasks are scheduled in a fair, simple round-robin fashion, where each task gets its turn. This is less optimal for very high concurrency situations compared to work-stealing.

- **Task**: Like Tokio, async-std’s tasks are future-based, but its goal is to provide a simpler model that feels familiar to Rust developers using synchronous `std` primitives.

#### 3. **smol**

- **Reactor**: Smol uses a tiny reactor based on the `polling` crate, which is lighter than `mio`. This reduces overhead, making it suitable for systems where low memory usage and simplicity are more critical than raw throughput.

- **Executor**: Smol provides a lightweight **single-threaded executor** by default, with optional multi-threading via the `async-executor` crate. It also supports a minimal work-stealing scheduler in a multi-threaded environment.

- **Task**: Smol's task model is designed to be minimal, with a focus on keeping task execution overhead very low. Tasks can be scheduled on a thread pool if needed, but the core is single-threaded for simplicity and reduced complexity.

### Advantages and Disadvantages

| Runtime      | **Advantages**                                               | **Disadvantages**                                               |
|--------------|---------------------------------------------------------------|-----------------------------------------------------------------|
| **Tokio**    | - High performance, excellent for high-concurrency workloads  <br> - Rich ecosystem, many utilities <br> - Highly scalable <br> - Multi-threaded or single-threaded modes | - More complex, steep learning curve <br> - Requires feature flags and configuration <br> - More heavyweight, higher memory usage |
| **async-std**| - Simpler API, more ergonomic <br> - Closer to synchronous Rust in feel <br> - Single-threaded by default, reducing complexity | - Lower performance in highly concurrent or intensive workloads <br> - Smaller ecosystem compared to Tokio |
| **smol**     | - Extremely lightweight and simple <br> - Minimal setup, no configuration <br> - Very low memory usage | - Limited features compared to Tokio <br> - Not as scalable for very high-concurrency environments <br> - Ecosystem is smaller |

### Summary

- **Tokio** is the go-to for high-performance applications that require extensive concurrency, networking, or complex I/O operations. Its work-stealing scheduler, reactor, and rich ecosystem make it ideal for large-scale systems.
- **async-std** is simpler and feels more like synchronous Rust, but trades some performance for ergonomics. It's suited for medium-scale workloads where ease of use is important.
- **smol** is lightweight, minimal, and best suited for smaller systems where low overhead and simplicity are key.

**Comment by anchalshivank** [comment]
Replica needs to be runtime agnostic, meaning it shouldn't be bound to a specific runtime. I believe this can be achieved. A relevant example is the [Ntex framework](https://github.com/ntex-rs/ntex?tab=readme-ov-file#usage), which provides flexibility for users to choose a runtime based on their needs. It demonstrates how a web framework can remain runtime-agnostic, which is something we can aim for with Replica.

**Comment by sergefdrv** [comment]
Ideally, we'd like to combine the simplicity of `smol` with scalable high-performance concurrency and parallelism of `tokio`.
